Add dedicated conventions documentation (#153)
EveronLife/EveronLife: .github/CONVENTIONS.md

```
src/
├─ Assets/
│  ├─ Resources/
│  │  ├─ Mining/
│  │  │  ├─ Data/
│  │  │  │  ├─ MetalOre_01_BCR.edds
│  │  │  │  ├─ MetalOre_01.emat
│  │  │  ├─ MetalOre_01.xob/
├─ Configs/
│  ├─ Feature/
│  │  ├─ Farming/
│  │  │  ├─ Mining/
│  │  │  │  ├─ OreSpawner.conf
├─ Prefabs/
│  ├─ Resources/
│  │  ├─ Mining/
│  │  │  ├─ MetalOre_01.et
├─ Scripts/
│  ├─ Feature/
│  │  ├─ Farming/
│  │  │  ├─ Mining/
│  │  │  │  ├─ EL_OreSpawner.c
│  │  │  │  ├─ EL_MineableComponent.c
│  │  │  │  ├─ UI/
│  │  │  │  │  ├─ MiningGui.c
├─ UI/
│  ├─ Layouts/
│  │  ├─ Feature/
│  │  │  ├─ Farming/
│  │  │  │  ├─ Mining/
│  │  │  │  │  ├─ MiningGui.layout
```

## <a name="code_guidelines"></a> Code Styleguide
As primary reference for how to write your code pelase see these:
- [Conventions](https://community.bistudio.com/wiki/Arma_Reforger:Scripting:_Conventions)
- [Do's and Don'ts](https://community.bistudio.com/wiki/Arma_Reforger:Scripting:_Do%27s_and_Don%27ts)
- [Performance](https://community.bistudio.com/wiki/Arma_Reforger:Scripting:_Performance)
- [Values](https://community.bistudio.com/wiki/Arma_Reforger:Scripting:_Values)

#### Flavour choices

> There are some "flavour" choices left for the individual developer. Please respect other peoples choices and do not reformat their code to use your flavor if you do not *need* to touch that piece of code fo the task at hand.

Reviewer: ```suggestion
There are some "flavour" choices left for the individual developer. Please respect other people's choices and do not reformat their code to use your flavor if you do not *need* to touch that piece of code fo the task at hand.
```

### ComponentEditorProps and EntityEditorProps
The attributes use to decorate `EntityClass` and `ComponentClass` classes follow the feature pattern described in the directory structure. Examples:
- `[EntityEditorProps(category: "EveronLife/Feature/Housing")]`
- `[ComponentEditorProps(category: "EveronLife/Feature/Farming/Mining")]`

### Modded class tagged names
When adding class members or functions throug a `modded` class include the TAG e.g. `EL` in the name to avoid conflicts. E.g. `m_iELModdedCount` or `void EL_Update()`.

### Single line statements
Single line statements that can be used on `if/else`, `for`, `foreach`, `while` are tolerated, but should be kept to a minimum to make the code readable and easy to debug/change.
Good use cases in include:
- `if (condition) return;`
- `if (condition) continue;`
- `if (instance) instance.Init();`

If the code becomes too complex, the line too long or the likely hood of wanting to debug something in there is reasonably high, put it into a code block with curly braces.

### Use of the auto keyword

> The keyword `auto` is only allowed if the right hand side expression is disclosing the expected compile time type.  

Reviewer: ```suggestion
The keyword `auto` is only allowed if the right hand side expression is disclosing the expected compile time type.  
```

> The keyword `auto` is only allowed if the right hand side expression is disclosing the expected compile time type.  

Reviewer: I dont really understand this.
I thought if you know the variable type you should not use auto and you know its "EL_SomeVeryLongComponentName" so why use auto?

Author: Because the line gets way too long and it is a lot to type. If the type is guaranteed by the right hand side and can be read without the need to go to the source and look it up, it is redundant information. Auto is meant for exactly this situation, same as `var` in c#.

Bad:
```
auto myVar = instance.GetValue(); 
```
Good:
```
auto myVeryLongComponent = EL_SomeVeryLongComponentName.Cast(entity.FindComponent(EL_SomeVeryLongComponentName));
```
**NOTE:** It does currently break autocompletion inside the Workbench so it is recommended (though not mandatory) to write out the type regardless.

## <a name="asset_guidelines"></a> Assets

### What kind of assets are added to the framework core?
The core of the framework will only ship assets that are needed for the minimal gameplay prototype. More variants or visual styles that do not fit with other existing assets are instead to be included within a seperate or even third party asset pack. Triple-check that the base game or the framework does not already have an asset that could be re-purposed with new materials/textures.

### Shared directories

> For some asset types there might be many of the same type. E.g. mutliple kinds of food like `Apple`, `Orange`, `Plum`. They are allowed to be put into one directory, then also sharing the same `Data` folder, as they are distinct by their names. If there is a logical grouping of many sub variants that belong together they can of course be put into a seperate sub folder that has its own `Data` folder then.

Reviewer: ```suggestion
For some asset types there might be many of the same type. E.g. multiple kinds of food like `Apple`, `Orange`, `Plum`. They are allowed to be put into one directory, then also sharing the same `Data` folder, as they are distinct by their names. If there is a logical grouping of many sub variants that belong together they can of course be put into a separate sub folder that has its own `Data` folder then.
```

### Data directory

> On import the game automatically creates a data folder for textures and materials used. Due to a [bug](https://feedback.bistudio.com/T165764) in the workbench the casing is incorrect. Always **manually** create a `Data` folder in the directory you want to import the object into.

Reviewer: ```suggestion
On import the game automatically creates a `data`folder for textures and materials used. Due to a [bug](https://feedback.bistudio.com/T165764) in the workbench the casing is incorrect. Always **manually** create a `Data` folder in the directory you want to import the object into.
```

- `MyObject_Base.et` (Add a base class to prefabs if it is likely that there will be many variants with a shared setup)
- `MyObject_01_Large_Blue.et` (Inherits from `MyObject_Base.et` is scaled to be big and has a blue material assigned.)

### File suffixes
Textures and materials need to be suffixed according to the slot they are made for e.g. `Name_BCR.edds`, `Name_NMO.edds`. The suffix serves a technical purpose for the game to select compression algorithms.  
Check out the vanilla game assets for reference and also read the [guide on texture creation](https://community.bistudio.com/wiki/Arma_Reforger:Textures). 
The [Arma Reforger Discord](https://discord.gg/arma) contains additional (often pinned) information. For textures (and materials) check out `#enfusion_texture_markers`
- [Texture compression explaination](https://discordapp.com/channels/105462288051380224/976231628785475624/981871081046569000)
- [Devs use TIFF with 8Bits and LZW](https://discordapp.com/channels/105462288051380224/976231628785475624/977200057357074432)
- [Substance painter export remarks](https://discordapp.com/channels/105462288051380224/976231628785475624/977154368526753812)

### File sizes

> One important factor in deciding if an asset will be included is its filesize. Because of the size limitations for console mods the framework needs to shave off every KB it can to give the individual servers as much room for addtional asset mods as possible. Metrics for triangle counts and texture sizes depend heavily on the pyhsical size and if it is a hand helt item or not.

Reviewer: ```suggestion
One important factor in deciding if an asset will be included is its file size. Because of the size limitations for console mods the framework needs to shave off every KB it can to give the individual servers as much room for additional asset mods as possible. Metrics for triangle counts and texture sizes depend heavily on the physical size and if it is a hand held item or not.
```
